cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,297 @@
+import io.cdap.cdap.common.io.DatumReader;
+import io.cdap.cdap.common.io.DatumWriter;
+import io.cdap.cdap.common.io.StringCachingDecoder;
+import io.cdap.cdap.common.utils.ResettableByteArrayInputStream;
+import io.cdap.cdap.internal.io.DatumReaderFactory;
+import io.cdap.cdap.internal.io.DatumWriterFactory;
+import io.cdap.cdap.internal.io.SchemaGenerator;
+import io.cdap.cdap.metrics.collect.AggregatedMetricsEmitter;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An aggregator for metrics that aggregates metrics on 2 levels:
+ * <ol>
+ *   <li>Combine metrics across all {@link MetricValues} with the same tags, retaining the latest timestamp.</li>
+ *   <li>Inside each {@link MetricValues}, combine {@link MetricValue} with the same name.</li>
+ * </ol>
+ */
+class MetricsMessageAggregator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      MetricsMessageAggregator.class);
+
+  private final Map<String, String> cachemap;
+  private final ResettableByteArrayInputStream decodingInputStream;
+  private final StringCachingDecoder decoder;
+  private final DatumReader<MetricValues> reader;
+  private final Schema schema;
+  private final ByteArrayOutputStream encodeOutputStream;
+  private final BinaryEncoder encoder;
+  private final DatumWriter<MetricValues> recordWriter;
+  private final long aggregationWindowSeconds;
+
+  MetricsMessageAggregator(SchemaGenerator schemaGenerator,
+      DatumWriterFactory writerFactory, DatumReaderFactory readerFactory,
+      long aggregationWindowSeconds) {
+    TypeToken<MetricValues> metricValueType = TypeToken.of(MetricValues.class);
+    try {
+      this.schema = schemaGenerator.generate(metricValueType.getType());
+    } catch (UnsupportedTypeException e) {
+      throw new RuntimeException("Failed to generate schema for MetricValues",
+          e);
+    }
+    this.reader = readerFactory.create(metricValueType, schema);
+    this.recordWriter = writerFactory.create(metricValueType, schema);
+    this.cachemap = new HashMap<>();
+    this.decodingInputStream = new ResettableByteArrayInputStream();
+    this.decoder = new StringCachingDecoder(
+        new BinaryDecoder(decodingInputStream), cachemap);
+    this.encodeOutputStream = new ByteArrayOutputStream(1024);
+    this.encoder = new BinaryEncoder(encodeOutputStream);
+    this.aggregationWindowSeconds = Math.max(1, aggregationWindowSeconds);
+  }
+
+  public Iterator<Message> aggregate(Iterator<Message> input) {
+    if (!input.hasNext()) {
+      return input;
+    }
+    cachemap.clear();
+
+    Map<MetricsMapKey, MetricsAggregator> aggregated = new HashMap<>();
+    String[] messageId = new String[1];
```

Comment:
Is this an array for performance reasons only?

Reply:
The java compiler doesn't allow mutation of objects that are captured inside a closure/lambda function. The objects need to be "effectively final", i.e have the final keyword or no mutations in the code that runs later. 
The object needs to be final, but its fields can be changed. Having array of size 1 tricks the compiler into thinking that the messageId isn't being mutated because the address of the array remains the same when we change its 0th element. The same trick is used [here](https://github.com/cdapio/cdap/blob/e81bb233fa54e757ace4bbd36a91eec48d967675/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java#L243).